Curly Braces Don't Work in Parametrized Text

Running my tests with version 25.0 to verify #269 it was found that curly braces in a parametrized text string are interpreted as formatting.
```
# test_bug.tavern.yaml
---

test_name: Bug Demonstration

marks:
  - bugdemo
  - parametrize:
      key: [line, text]
      vals:
        # NOTE: "\" requires doubling, !raw will take care of "{" and "}"
        - [1, "XYZ[\\]^_`abcdefghijk"]
        - [2, !raw "lmnopqrstuvwxyz{|}~*"]

paho-mqtt: *mqtt_spec

stages:
  - name: Step 1 - Display Line
    mqtt_publish:
      topic: dcp/target_display_line
      json:
        line_no: !int "{line}"
        line_text: "{text}"
    mqtt_response:
      topic: dcp/dcp_response
      json:
        result_code: !int 0
``` 
yields an error -
```
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   tavern.util.exceptions.MissingFormatError: ('|',)

<string>:1: MissingFormatError
------------------------------ Captured log call -------------------------------
dict_util.py                41 ERROR    Failed to resolve string [lmnopqrstuvwxyz{|}~] with variables [{'tavern': {'env_vars': {'XDG_VTNR': '7', 'XDG_SESSION_ID': 'c2', 'XDG_GREETER_DATA_DIR': '/var/lib/lightdm-data/vagrant', 'SELINUX_INIT': 'YES', 'CLUTTER_IM_MODULE': 'xim', 'GPG_AGENT_INFO': '/run/user/1000/keyring-kypD5L/gpg:0:1', 'TERM': 'xterm-256color', 'VTE_VERSION': '3409', 'SHELL': '/bin/bash', 'DERBY_HOME': '/usr/lib/jvm/java-8-oracle/db', 'WINDOWID': '69213036', 'OLDPWD': '/media/sf_GIT_Repository/TDCPSimulator/temp', 'UPSTART_SESSION': 'unix:abstract=/com/ubuntu/upstart-session/1000/2180', 'GNOME_KEYRING_CONTROL': '/run/user/1000/keyring-kypD5L', 'GTK_MODULES': 'overlay-scrollbar:unity-gtk-module', 'NO_PROXY': 'localhost,127.0.0.1', 'http_proxy': 'http://194.145.60.1:9400', 'USER': 'vagrant', 'LS_COLORS': 'rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arj=01;31:*.taz=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.zip=01;31:*.z=01;31:*.Z=01;31:*.dz=01;31:*.gz=01;31:*.lz=01;31:*.xz=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.jpg=01;35:*.jpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.axv=01;35:*.anx=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.axa=00;36:*.oga=00;36:*.spx=00;36:*.xspf=00;36:', 'LD_LIBRARY_PATH': '/usr/local/lib', 'XDG_SESSION_PATH': '/org/freedesktop/DisplayManager/Session0', 'XDG_SEAT_PATH': '/org/freedesktop/DisplayManager/Seat0', 'SSH_AUTH_SOCK': '/run/user/1000/keyring-kypD5L/ssh', 'FTP_PROXY': '', 'ftp_proxy': '', 'DEFAULTS_PATH': '/usr/share/gconf/ubuntu.default.path', 'XDG_CONFIG_DIRS': '/etc/xdg/xdg-ubuntu:/usr/share/upstart/xdg:/etc/xdg', 'all_proxy': 'socks://.:8080/', 'ALL_PROXY': 'socks://.:8080/', 'DESKTOP_SESSION': 'ubuntu', 'PATH': '/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/usr/lib/jvm/java-8-oracle/bin:/usr/lib/jvm/java-8-oracle/db/bin:/usr/lib/jvm/java-8-oracle/jre/bin', 'QT_IM_MODULE': 'ibus', 'QT_QPA_PLATFORMTHEME': 'appmenu-qt5', 'PWD': '/media/sf_GIT_Repository/TDCPSimulator/Tests', 'JOB': 'gnome-session', 'XMODIFIERS': '@im=ibus', 'JAVA_HOME': '/usr/lib/jvm/java-8-oracle', 'GNOME_KEYRING_PID': '2167', 'LANG': 'en_US.UTF-8', 'GDM_LANG': 'en', 'MANDATORY_PATH': '/usr/share/gconf/ubuntu.mandatory.path', 'COMPIZ_CONFIG_PROFILE': 'ubuntu', 'IM_CONFIG_PHASE': '1', 'https_proxy': 'https://194.145.60.1:9400', 'GDMSESSION': 'ubuntu', 'HTTPS_PROXY': 'https://194.145.60.1:9400', 'SESSIONTYPE': 'gnome-session', 'SHLVL': '1', 'HOME': '/home/vagrant', 'XDG_SEAT': 'seat0', 'LANGUAGE': 'en', 'LIBGL_ALWAYS_SOFTWARE': '1', 'no_proxy': 'localhost,127.0.0.1', 'GNOME_DESKTOP_SESSION_ID': 'this-is-deprecated', 'HTTP_PROXY': 'http://194.145.60.1:9400', 'UPSTART_INSTANCE': '', 'UPSTART_EVENTS': 'started starting', 'LOGNAME': 'vagrant', 'DBUS_SESSION_BUS_ADDRESS': 'unix:abstract=/tmp/dbus-yP9DmMYUld', 'J2SDKDIR': '/usr/lib/jvm/java-8-oracle', 'XDG_DATA_DIRS': '/usr/share/ubuntu:/usr/share/gnome:/usr/local/share/:/usr/share/', 'QT4_IM_MODULE': 'xim', 'LESSOPEN': '| /usr/bin/lesspipe %s', 'INSTANCE': 'Unity', 'TEXTDOMAIN': 'im-config', 'UPSTART_JOB': 'unity-settings-daemon', 'XDG_RUNTIME_DIR': '/run/user/1000', 'DISPLAY': ':0', 'XDG_CURRENT_DESKTOP': 'Unity', 'GTK_IM_MODULE': 'ibus', 'J2REDIR': '/usr/lib/jvm/java-8-oracle/jre', 'LESSCLOSE': '/usr/bin/lesspipe %s %s', 'TEXTDOMAINDIR': '/usr/share/locale/', 'XAUTHORITY': '/home/vagrant/.Xauthority', 'COLORTERM': 'gnome-terminal', '_': '/usr/local/bin/py.test', 'PYTEST_CURRENT_TEST': 'test_unicode.tavern.yaml::Display Line - Unicode[1-lmnopqrstuvwxyz{|}~\x7f-lmnopqrstuvwxyz{?}?\x7f] (call)'}}}]
dict_util.py                43 ERROR    Key(s) not found in format: ('|',)
============= 1 failed, 1 passed, 64 deselected in 1.22 seconds ==============
```
As with #269 the test worked correctly if the data was explicitly placed in the test rather than parametrized.

Note: Using single quotes (without character doubling) produces a similar error, but in the collection phase.
```
------------------------------- Captured stderr --------------------------------
Failed to resolve string [lmnopqrstuvwxyz{|}~*] with variables [{}]
Key(s) not found in format: ('|',)
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!
```
Unlike #269 this was performed using Python 3.6.

--- a/tavern/util/dict_util.py
+++ b/tavern/util/dict_util.py
@@ -13,7 +13,7 @@ from . import exceptions
 logger = logging.getLogger(__name__)
 
 
-class _FormattedString(str):
+class _FormattedString(object):
     """Wrapper class for things that have already been formatted
 
     This is only used below and should not be used outside this module
@@ -58,7 +58,10 @@ def format_keys(val, variables):
             logger.error("Empty format values are invalid")
             raise_from(exceptions.MissingFormatError(e.args), e)
 
-        formatted = _FormattedString(formatted)
+        class InnerFormattedString(_FormattedString, type(val)):
+            """Hack for python 2"""
+
+        formatted = InnerFormattedString(formatted)
     elif isinstance(val, TypeConvertToken):
         value = format_keys(val.value, box_vars)
         formatted = val.constructor(value)

--- a/tavern/util/dict_util.py
+++ b/tavern/util/dict_util.py
@@ -13,6 +13,16 @@ from . import exceptions
 logger = logging.getLogger(__name__)
 
 
+class _FormattedString(str):
+    """Wrapper class for things that have already been formatted
+
+    This is only used below and should not be used outside this module
+    """
+
+    def __init(self, s):
+        super(_FormattedString, self).__init__(s)
+
+
 def format_keys(val, variables):
     """recursively format a dictionary with the given values
 
@@ -33,6 +43,8 @@ def format_keys(val, variables):
             formatted[key] = format_keys(val[key], box_vars)
     elif isinstance(val, (list, tuple)):
         formatted = [format_keys(item, box_vars) for item in val]
+    elif isinstance(formatted, _FormattedString):
+        logger.debug("Already formatted %s, not double-formatting", formatted)
     elif isinstance(val, (ustr, str)):
         try:
             formatted = val.format(**box_vars)
@@ -45,9 +57,13 @@ def format_keys(val, variables):
         except IndexError as e:
             logger.error("Empty format values are invalid")
             raise_from(exceptions.MissingFormatError(e.args), e)
+
+        formatted = _FormattedString(formatted)
     elif isinstance(val, TypeConvertToken):
         value = format_keys(val.value, box_vars)
         formatted = val.constructor(value)
+    else:
+        logger.debug("Not formatting something of type '%s'", type(formatted))
 
     return formatted
 

--- a/tests/integration/test_parametrize.tavern.yaml
+++ b/tests/integration/test_parametrize.tavern.yaml
@@ -470,3 +470,29 @@ stages:
       status_code: 200
       body:
         value: "{generic_str}-{edible}"
+
+---
+
+test_name: Test that double formatting something in
+
+marks:
+  - parametrize:
+      key:
+        - line
+        - text
+      vals:
+        # NOTE: "\" requires doubling, !raw will take care of "{" and "}"
+        - [1, "XYZ[\\]^_`abcdefghijk"]
+        - [2, !raw "lmnopqrstuvwxyz{|}~*"]
+
+stages:
+  - name: Echo back parametrized text
+    request:
+      url: "{global_host}/echo"
+      method: POST
+      json:
+        value: "{line}-{text}"
+    response:
+      status_code: 200
+      body:
+        value: "{line}-{text}"

--- a/tests/unit/test_utilities.py
+++ b/tests/unit/test_utilities.py
@@ -1,7 +1,4 @@
 import contextlib
-import os
-import tempfile
-from textwrap import dedent
 import copy
 import os
 import tempfile
@@ -10,13 +7,11 @@ from textwrap import dedent
 
 import pytest
 import yaml
-
 from mock import Mock, patch
 
 from tavern.schemas.extensions import validate_extensions
 from tavern.schemas.files import wrapfile
 from tavern.util import exceptions
-from tavern.util.loader import ANYTHING, IncludeLoader, construct_include
 from tavern.util.dict_util import (
     deep_dict_merge,
     check_keys_match_recursive,
@@ -24,6 +19,7 @@ from tavern.util.dict_util import (
     recurse_access_key,
 )
 from tavern.util.loader import ANYTHING, IncludeLoader, load_single_document_yaml
+from tavern.util.loader import construct_include
 
 
 class TestValidateFunctions:
@@ -240,6 +236,18 @@ class TestFormatKeys:
         format_variables = {"b": final_value}
 
         assert format_keys(to_format, format_variables)["a"] == final_value
+
+    def test_no_double_format_failure(self):
+        to_format = "{{b}}"
+
+        final_value = "{b}"
+
+        format_variables = {"b": final_value}
+
+        formatted = format_keys(to_format, format_variables)
+        assert formatted == final_value
+        formatted_2 = format_keys(formatted, {})
+        assert formatted_2 == final_value
 
 
 class TestRecurseAccess:

--- a/tox-integration.ini
+++ b/tox-integration.ini
@@ -1,5 +1,5 @@
 [tox]
-envlist = {py27,py34,py35,py36,py37,pypy,pypy3}-{generic,cookies,mqtt,advanced,components}
+envlist = {py27,py34,py35,py36,py37,pypy,pypy3}-{generic,cookies,mqtt,advanced,components,noextra}
 skip_missing_interpreters = true
 
 [testenv]
@@ -14,6 +14,7 @@ changedir =
     advanced: example/advanced
     components: example/components
     generic: tests/integration
+    noextra: tests/integration
 deps =
     docker-compose
     pytest>=3.6

